File: room/integration-tests/testapp/src/androidTest/java/androidx/room/integration/testapp/test/NoBackupDirectoryTest.java
/*
 * Copyright 2019 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.room.integration.testapp.test;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.notNullValue;
import static org.hamcrest.MatcherAssert.assertThat;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import androidx.room.integration.testapp.TestDatabase;
import androidx.sqlite.db.SupportSQLiteDatabase;
import androidx.sqlite.db.SupportSQLiteOpenHelper;
import androidx.sqlite.db.framework.FrameworkSQLiteOpenHelperFactory;
import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.test.filters.SdkSuppress;
import androidx.test.filters.SmallTest;
import androidx.test.platform.app.InstrumentationRegistry;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.io.File;

@RunWith(AndroidJUnit4.class)
@SmallTest
public class NoBackupDirectoryTest {

    private static final String NAME = "database.db";

    private Context mContext;
    private SupportSQLiteOpenHelper.Factory mFactory;

    @Before
    public void setUp() {
        mContext = InstrumentationRegistry.getInstrumentation().getTargetContext();
        mFactory = new AlwaysNoBackupFactory();
    }

    @Test
    @SdkSuppress(minSdkVersion = 23)
    public void testBuildDatabase_withCustomSupportSQLiteOpenHelper() {
        // Setting the minSdkVersion to 23 even though Context.getNoBackupFilesDir() is supported
        // on API 21+ because it was unused until API 23.
        RoomDatabase database = Room.databaseBuilder(mContext, TestDatabase.class, NAME)
                .openHelperFactory(mFactory)
                .build();

        SupportSQLiteOpenHelper helper = database.getOpenHelper();
        assertThat(helper.getDatabaseName(), is(NAME));
        SupportSQLiteDatabase sqLiteDatabase = helper.getWritableDatabase();
        String expectedPath = new File(mContext.getNoBackupFilesDir(), NAME).getPath();
        assertThat(sqLiteDatabase.getPath(), is(expectedPath));
        database.close();
    }

    @Test
    public void testBuildInMemoryDatabase_withAlwaysNoBackupFactory() {
        Throwable exception = null;
        try {
            RoomDatabase database = Room.inMemoryDatabaseBuilder(mContext, TestDatabase.class)
                    .openHelperFactory(mFactory)
                    .build();
            database.getOpenHelper().getWritableDatabase();
        } catch (Throwable expected) {
            exception = expected;
        }

        assertThat(exception, notNullValue());
        assertThat(exception.getMessage(),
                is("Must set a non-null database name to a configuration that uses the no backup "
                        + "directory."));
    }

    static class AlwaysNoBackupFactory implements SupportSQLiteOpenHelper.Factory {
        private final SupportSQLiteOpenHelper.Factory mDelegate;

        AlwaysNoBackupFactory() {
            mDelegate = new FrameworkSQLiteOpenHelperFactory();
        }

        @NonNull
        @Override
        public SupportSQLiteOpenHelper create(
                @NonNull SupportSQLiteOpenHelper.Configuration configuration) {
            SupportSQLiteOpenHelper.Configuration backupConfiguration =
                    SupportSQLiteOpenHelper.Configuration.builder(configuration.context)
                            .callback(configuration.callback)
                            .name(configuration.name)
                            .noBackupDirectory(true)
                            .build();
            return mDelegate.create(backupConfiguration);
        }
    }
}

File: room3/integration-tests/kotlintestapp/src/androidTest/java/androidx/room3/integration/kotlintestapp/test/NoBackupDirectoryTest.kt
/*
 * Copyright 2025 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.room3.integration.kotlintestapp.test

import android.os.StrictMode
import android.os.StrictMode.ThreadPolicy
import androidx.kruth.assertWithMessage
import androidx.room3.Room
import androidx.room3.integration.kotlintestapp.TestDatabase
import androidx.room3.integration.kotlintestapp.test.TestDatabaseTest.UseDriver
import androidx.sqlite.SQLiteConnection
import androidx.sqlite.SQLiteDriver
import androidx.sqlite.driver.AndroidSQLiteDriver
import androidx.sqlite.driver.bundled.BundledSQLiteDriver
import androidx.test.filters.SmallTest
import androidx.test.platform.app.InstrumentationRegistry
import java.io.File
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.test.runTest
import kotlinx.coroutines.withContext
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized
import org.junit.runners.Parameterized.Parameters

@SmallTest
@RunWith(Parameterized::class)
class NoBackupDirectoryTest(private val useDriver: UseDriver) {

    private companion object {
        @JvmStatic
        @Parameters(name = "useDriver={0}")
        fun parameters() = arrayOf(UseDriver.ANDROID, UseDriver.BUNDLED)
    }

    private val context = InstrumentationRegistry.getInstrumentation().targetContext

    @Test
    fun testNoBackupDirectoryFileName() = runTest {
        val databaseFile = File(context.noBackupFilesDir, "temporal.db")

        withContext(Dispatchers.Main) {
            // This is an easy way to use the no backup dir, but the API getNoBackupFilesDir() does
            // IO and depending where the Room builder is created, it can be in the main thread.
            val database =
                Room.databaseBuilder<TestDatabase>(context, databaseFile.path)
                    .setDriver(
                        when (useDriver) {
                            UseDriver.ANDROID -> AndroidSQLiteDriver()
                            UseDriver.BUNDLED -> BundledSQLiteDriver()
                        }
                    )
                    .build()
            database.booksDao().insertPublisherSuspend("p1", "pub1")
            database.close()
        }

        assertWithMessage("Expected the database file in the no backup dir to be created.")
            .that(databaseFile.exists())
            .isTrue()
        databaseFile.delete()
    }

    @Test
    fun testNoBackupDirectoryDriver() = runTest {
        // Enable StrictMode to validate no IO is done in the main thread
        val currentPolicy = StrictMode.getThreadPolicy()
        StrictMode.setThreadPolicy(
            ThreadPolicy.Builder().detectDiskReads().detectDiskWrites().penaltyDeath().build()
        )

        // This is a more appropriate way to use the no backup dir, as the driver open() call is
        // done off the main thread and will be OK with the getNoBackupFilesDir() IO.
        val actualDriver =
            when (useDriver) {
                UseDriver.ANDROID -> AndroidSQLiteDriver()
                UseDriver.BUNDLED -> BundledSQLiteDriver()
            }
        val noBackupDirDriver =
            object : SQLiteDriver by actualDriver {
                override fun open(fileName: String): SQLiteConnection {
                    val noBackupFileName =
                        File(context.noBackupFilesDir, fileName.substringAfterLast("/"))
                    return actualDriver.open(noBackupFileName.path)
                }
            }
        withContext(Dispatchers.Main) {
            val database =
                Room.databaseBuilder<TestDatabase>(context, "temporal.db")
                    .setDriver(noBackupDirDriver)
                    .build()
            database.booksDao().insertPublisherSuspend("p1", "pub1")
            database.close()
        }

        StrictMode.setThreadPolicy(currentPolicy)

        val databaseFile = File(context.noBackupFilesDir, "temporal.db")
        assertWithMessage("Expected the database file in the no backup dir to be created.")
            .that(databaseFile.exists())
            .isTrue()
        databaseFile.delete()
    }
}
